lib/stdio/generator.js

```diff
@@ -0,0 +1,86 @@
+import {Duplex, Readable, PassThrough, getDefaultHighWaterMark} from 'node:stream';
+import {callbackify} from 'node:util';
+import {setTimeout} from 'node:timers/promises';
+
+// Generators can be used to transform/filter standard streams.
+// Generators have a simple syntax, yet allows all of the following:
+//  - Sharing state between chunks, by using logic before the `for` loop
+//  - Flushing logic, by using logic after the `for` loop
+//  - Asynchronous logic
+//  - Emitting multiple chunks from a single source chunk, even if spaced in time, by using multiple `yield`
+//  - Filtering, by using no `yield`
+// Therefore, there is no need to allow Node.js or web transform streams.
+// The `highWaterMark` is kept as the default value, since this is what `childProcess.std*` uses.
+// We ensure `objectMode` is `false` for better buffering.
+// Chunks are currently processed serially. We could add a `concurrency` option to parallelize in the future.
+export const generatorToTransformStream = ({value}, {encoding}) => {
+	const objectMode = false;
+	const highWaterMark = getDefaultHighWaterMark(objectMode);
+	const passThrough = new PassThrough({objectMode, highWaterMark, destroy: destroyPassThrough});
+	const iterable = passThrough.iterator();
+	const encodedIterable = applyEncoding(iterable, encoding);
+	const mappedIterable = value(encodedIterable);
+	const readableStream = Readable.from(mappedIterable, {objectMode, highWaterMark});
+	const duplexStream = Duplex.from({writable: passThrough, readable: readableStream});
+	return {value: duplexStream};
+};
```

Comment:
Description of what's going in here, because it is rather complex.

We return a `Duplex`, created by `Duplex.from()` made of a writable stream and a readable stream, piped to each other.

The writable stream is a simple `PassThrough`, so it does not do anything except forwarding data to the readable part.

The `PassThrough` is read as an iterable using `passThrough.iterator()`. 

This iterable is transformed to another iterable, by applying the encoding generators. Those convert the chunk type from `Buffer` to either `string` or `Uint8Array` depending on the `encoding` option.

This new iterable is transformed again to another one, this time by applying the user-supplied generator.

Finally, `Readable.from()` is used to convert this final iterable to a `Readable` stream.

I have carefully checked the following points, and added tests for it:
  - Error handling: the right errors are propagated. Streams end/abort/errors are propagated to other streams. Streams are properly destroyed.
  - Buffering with `highWaterMark`

Reply:
The description here would make a code code comment for future readers of the code.

Reply:
Done in 71055df5c889f0370e55de9eb397529bb0bfb0cf :+1: